invokeai/backend/stable_diffusion/extensions/freeu.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import annotations
+
+from contextlib import contextmanager
+from typing import TYPE_CHECKING, Dict, Optional
+
+import torch
+from diffusers import UNet2DConditionModel
+
+from invokeai.backend.stable_diffusion.extensions.base import ExtensionBase
+
+if TYPE_CHECKING:
+    from invokeai.app.shared.models import FreeUConfig
+
+
+class FreeUExt(ExtensionBase):
+    def __init__(
+        self,
+        freeu_config: Optional[FreeUConfig],
+    ):
+        super().__init__()
+        self.freeu_config = freeu_config
+
+    @contextmanager
+    def patch_unet(self, unet: UNet2DConditionModel, cached_weights: Optional[Dict[str, torch.Tensor]] = None):
+        did_apply_freeu = False
+        try:
+            assert hasattr(unet, "enable_freeu")  # mypy doesn't pick up this attribute?
+            if self.freeu_config is not None:
+                unet.enable_freeu(
+                    b1=self.freeu_config.b1,
+                    b2=self.freeu_config.b2,
+                    s1=self.freeu_config.s1,
+                    s2=self.freeu_config.s2,
+                )
+                did_apply_freeu = True
+
+            yield
+
+        finally:
+            assert hasattr(unet, "disable_freeu")  # mypy doesn't pick up this attribute?
+            if did_apply_freeu:
+                unet.disable_freeu()
```

Comment:
This code block can be simplified in the following ways:
-  "mypy doesn't pick up this attribute?" -> I'm not sure what was causing this problem, but I tested and both pyright and mypy have no issue inferring the expected methods.
- `if self.freeu_config is not None:` can be removed as noted on an earlier comment.
- There is no need for `did_apply_freeu`. By moving `enable_freeu(...)` before the `try` block, we know that we will only enter the `try-finally` if `enable_freeu(...)` was applied. Also, note that there was no `catch` clause so this does not change the exception-throwing behavior at all.
```suggestion
        unet.enable_freeu(
            b1=self._freeu_config.b1,
            b2=self._freeu_config.b2,
            s1=self._freeu_config.s1,
            s2=self._freeu_config.s2,
        )
        try:
            yield
        finally:
            unet.disable_freeu()
```